From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Gr1mly4Memes <alagozantoni14@gmail.com>
Date: Sat, 17 Jan 2026 19:08:20 -0800
Subject: [PATCH] Leaves Lava Riptide


diff --git a/net/minecraft/world/item/TridentItem.java b/net/minecraft/world/item/TridentItem.java
index 4f63ebf5ff1381fc69d4d6e73d326fe6833e3654..b7d8a041affd095364181368988be3f6322577da 100644
--- a/net/minecraft/world/item/TridentItem.java
+++ b/net/minecraft/world/item/TridentItem.java
@@ -70,7 +70,7 @@ public class TridentItem extends Item implements ProjectileItem {
                 return false;
             } else {
                 float tridentSpinAttackStrength = EnchantmentHelper.getTridentSpinAttackStrength(stack, player);
-                if (tridentSpinAttackStrength > 0.0F && !player.isInWaterOrRain()) {
+                if (tridentSpinAttackStrength > 0.0F && !player.isInWaterOrRain() || (gr1mly4memes.nitro.NitroConfig.lavaRiptide && player.isInLava())) { // Nitro //Leaves - lava riptide
                     return false;
                 } else if (stack.nextDamageWillBreak()) {
                     return false;
@@ -161,7 +161,7 @@ public class TridentItem extends Item implements ProjectileItem {
         ItemStack itemInHand = player.getItemInHand(hand);
         if (itemInHand.nextDamageWillBreak()) {
             return InteractionResult.FAIL;
-        } else if (EnchantmentHelper.getTridentSpinAttackStrength(itemInHand, player) > 0.0F && !player.isInWaterOrRain()) {
+        } else if (EnchantmentHelper.getTridentSpinAttackStrength(itemInHand, player) > 0.0F && !player.isInWaterOrRain() && !(gr1mly4memes.nitro.NitroConfig.lavaRiptide && player.isInLava())) { // Nitro //Leaves - lava riptide
             return InteractionResult.FAIL;
         } else {
             player.startUsingItem(hand);
