From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Gr1mly4Memes <alagozantoni14@gmail.com>
Date: Thu, 8 Jan 2026 19:00:08 -0800
Subject: [PATCH] lithium: combined_heightmap_update


diff --git a/net/minecraft/world/level/chunk/LevelChunk.java b/net/minecraft/world/level/chunk/LevelChunk.java
index 54802b358b5e6d0bfde09713de0ddb20f015ffd5..43779c9049d9fc08f37b62ee23c312d4ac936a78 100644
--- a/net/minecraft/world/level/chunk/LevelChunk.java
+++ b/net/minecraft/world/level/chunk/LevelChunk.java
@@ -392,10 +392,13 @@ public class LevelChunk extends ChunkAccess implements DebugValueSource, ca.spot
                 return null;
             } else {
                 Block block = state.getBlock();
-                this.heightmaps.get(Heightmap.Types.MOTION_BLOCKING).update(i, y, i2, state);
-                this.heightmaps.get(Heightmap.Types.MOTION_BLOCKING_NO_LEAVES).update(i, y, i2, state);
-                this.heightmaps.get(Heightmap.Types.OCEAN_FLOOR).update(i, y, i2, state);
-                this.heightmaps.get(Heightmap.Types.WORLD_SURFACE).update(i, y, i2, state);
+                // Nitro start - lithium: combined_heightmap_update
+                Heightmap heightmap0 = this.heightmaps.get(Heightmap.Types.MOTION_BLOCKING);
+                Heightmap heightmap1 = this.heightmaps.get(Heightmap.Types.MOTION_BLOCKING_NO_LEAVES);
+                Heightmap heightmap2 = this.heightmaps.get(Heightmap.Types.OCEAN_FLOOR);
+                Heightmap heightmap3 = this.heightmaps.get(Heightmap.Types.WORLD_SURFACE);
+                net.caffeinemc.mods.lithium.common.world.chunk.heightmap.CombinedHeightmapUpdate.updateHeightmaps(heightmap0, heightmap1, heightmap2, heightmap3, this, i, y, i2, state);
+                // Nitro end - lithium: combined_heightmap_update
                 boolean hasOnlyAir1 = section.hasOnlyAir();
                 if (hasOnlyAir != hasOnlyAir1) {
                     this.level.getChunkSource().getLightEngine().updateSectionStatus(pos, hasOnlyAir1);
diff --git a/net/minecraft/world/level/levelgen/Heightmap.java b/net/minecraft/world/level/levelgen/Heightmap.java
index 25c537496ccea6fe27111b21fe2e47b3f6a914a1..1247f67376067d5054078cc2721f238cc6871471 100644
--- a/net/minecraft/world/level/levelgen/Heightmap.java
+++ b/net/minecraft/world/level/levelgen/Heightmap.java
@@ -123,6 +123,12 @@ public class Heightmap {
         this.data.set(getIndex(x, z), value - this.chunk.getMinY());
     }
 
+    // Nitro start - lithium: combined_heightmap_update
+    public final Predicate<BlockState> isOpaque() {
+        return this.isOpaque;
+    }
+    // Nitro end - lithium: combined_heightmap_update
+
     public void setRawData(ChunkAccess chunk, Heightmap.Types type, long[] data) {
         long[] raw = this.data.getRaw();
         if (raw.length == data.length) {
