From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Gr1mly4Memes <alagozantoni14@gmail.com>
Date: Mon, 2 Feb 2026 19:32:56 -0800
Subject: [PATCH] amplifiedWorld


diff --git a/net/minecraft/server/dedicated/DedicatedServer.java b/net/minecraft/server/dedicated/DedicatedServer.java
index f9f1d09335956e6a0081356df7bcd53dc07f19ad..8af6182e887c738828ae2d7150885fc92c30f784 100644
--- a/net/minecraft/server/dedicated/DedicatedServer.java
+++ b/net/minecraft/server/dedicated/DedicatedServer.java
@@ -410,6 +410,11 @@ public class DedicatedServer extends MinecraftServer implements ServerInterface
             long nanos = Util.getNanos();
             this.services.nameToIdCache().resolveOfflineUsers(!this.usesAuthentication());
             LOGGER.info("Preparing level \"{}\"", this.getLevelIdName());
+            // Nitro start - Amplified World
+            if (gr1mly4memes.nitro.NitroConfig.amplifiedWorld) {
+                this.worldData.worldGenOptions().withAmplified(true);
+            }
+            // Nitro end - Amplified World
             this.loadLevel(this.storageSource.getLevelId()); // CraftBukkit
             long l = Util.getNanos() - nanos;
             String string3 = String.format(Locale.ROOT, "%.3fs", l / 1.0E9);
diff --git a/net/minecraft/server/dedicated/DedicatedServerProperties.java b/net/minecraft/server/dedicated/DedicatedServerProperties.java
index 2a3b10c5e031ef44a2b53bc4fdc6e86c917f75d4..61f1dda4fca28e2142623fdd2d4c2214ee991a23 100644
--- a/net/minecraft/server/dedicated/DedicatedServerProperties.java
+++ b/net/minecraft/server/dedicated/DedicatedServerProperties.java
@@ -266,6 +266,11 @@ public class DedicatedServerProperties extends Settings<DedicatedServerPropertie
             Holder.Reference<WorldPreset> reference = holderLookup.get(WorldPresets.NORMAL)
                 .or(() -> holderLookup.listElements().findAny())
                 .orElseThrow(() -> new IllegalStateException("Invalid datapack contents: can't find default preset"));
+            // Nitro start - Amplified World
+            if (gr1mly4memes.nitro.NitroConfig.amplifiedWorld) {
+                return holderLookup.get(WorldPresets.AMPLIFIED).orElse(reference).value().createWorldDimensions();
+            }
+            // Nitro end - Amplified World
             Holder<WorldPreset> holder = Optional.ofNullable(Identifier.tryParse(this.levelType))
                 .map(presetPath -> ResourceKey.create(Registries.WORLD_PRESET, presetPath))
                 .or(() -> Optional.ofNullable(LEGACY_PRESET_NAMES.get(this.levelType)))
diff --git a/net/minecraft/world/level/levelgen/WorldOptions.java b/net/minecraft/world/level/levelgen/WorldOptions.java
index c92508741439a8d0d833ea02d0104416adb83c92..aaee83ff916f0bb8a36ed2ac03ab97e12115b6c2 100644
--- a/net/minecraft/world/level/levelgen/WorldOptions.java
+++ b/net/minecraft/world/level/levelgen/WorldOptions.java
@@ -23,6 +23,7 @@ public class WorldOptions {
     private final boolean generateStructures;
     private final boolean generateBonusChest;
     private final Optional<String> legacyCustomOptions;
+    private boolean amplified; // Nitro - Amplified World
 
     public WorldOptions(long seed, boolean generateStructures, boolean generateBonusChest) {
         this(seed, generateStructures, generateBonusChest, Optional.empty());
@@ -87,4 +88,14 @@ public class WorldOptions {
     public static long randomSeed() {
         return RandomSource.create().nextLong();
     }
+
+    // Nitro start - Amplified World
+    public void withAmplified(boolean amplified) {
+        this.amplified = amplified;
+    }
+
+    public boolean isAmplified() {
+        return this.amplified;
+    }
+    // Nitro end - Amplified World
 }
