From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Gr1mly4Memes <alagozantoni14@gmail.com>
Date: Fri, 6 Feb 2026 19:15:04 -0800
Subject: [PATCH] PulseMC Features


diff --git a/net/minecraft/commands/Commands.java b/net/minecraft/commands/Commands.java
index 83148f756ecc77ae45abffbd1d7cbfacf6c99e7f..e951fb971c1ffbdd9206f5ba5b736e0f4b9f3679 100644
--- a/net/minecraft/commands/Commands.java
+++ b/net/minecraft/commands/Commands.java
@@ -261,6 +261,7 @@ public class Commands {
         WaypointCommand.register(this.dispatcher, context);
         WeatherCommand.register(this.dispatcher);
         WorldBorderCommand.register(this.dispatcher);
+        dev.pulsemc.pulse.command.PulseCommands.register(this.dispatcher); // Pulse - Register main command
         if (JvmProfiler.INSTANCE.isAvailable()) {
             JfrCommand.register(this.dispatcher);
         }
diff --git a/net/minecraft/network/protocol/game/ClientboundSectionBlocksUpdatePacket.java b/net/minecraft/network/protocol/game/ClientboundSectionBlocksUpdatePacket.java
index cb0dd1aa58b8b5972cd8410e1e22fda64f329105..233dabacebecd931ae722383bcbd1b32690826ed 100644
--- a/net/minecraft/network/protocol/game/ClientboundSectionBlocksUpdatePacket.java
+++ b/net/minecraft/network/protocol/game/ClientboundSectionBlocksUpdatePacket.java
@@ -17,9 +17,11 @@ public class ClientboundSectionBlocksUpdatePacket implements Packet<ClientGamePa
         ClientboundSectionBlocksUpdatePacket::write, ClientboundSectionBlocksUpdatePacket::new
     );
     private static final int POS_IN_SECTION_BITS = 12;
-    private final SectionPos sectionPos;
-    private final short[] positions;
-    private final BlockState[] states;
+    // Pulse start - public for PulseBuffer
+    public final SectionPos sectionPos;
+    public final short[] positions;
+    public final BlockState[] states;
+    // Pulse end - public for PulseBuffer
 
     public ClientboundSectionBlocksUpdatePacket(SectionPos sectionPos, ShortSet positions, LevelChunkSection section) {
         this.sectionPos = sectionPos;
diff --git a/net/minecraft/server/dedicated/DedicatedServer.java b/net/minecraft/server/dedicated/DedicatedServer.java
index f9f1d09335956e6a0081356df7bcd53dc07f19ad..4e8097107f4e68f2f340a01063bfcbd43cb2a3b6 100644
--- a/net/minecraft/server/dedicated/DedicatedServer.java
+++ b/net/minecraft/server/dedicated/DedicatedServer.java
@@ -368,6 +368,12 @@ public class DedicatedServer extends MinecraftServer implements ServerInterface
         }
         // Purpur end - UPnP Port Forwarding
 
+        // Pulse start - Init Pulse systems
+        dev.pulsemc.pulse.ConfigManager.load();
+        dev.pulsemc.pulse.metrics.Metrics.start();
+        dev.pulsemc.pulse.metrics.extensions.MetricsBar.start();
+        // Pulse end - Init Pulse systems
+
         // CraftBukkit start
         this.server.loadPlugins();
         this.server.enablePlugins(org.bukkit.plugin.PluginLoadOrder.STARTUP);
diff --git a/net/minecraft/server/network/ServerCommonPacketListenerImpl.java b/net/minecraft/server/network/ServerCommonPacketListenerImpl.java
index d5bf3b49d34a04294c92a5b5b5d2de290642fc5e..3e18868fd0be973d068ab265fbd6575bd1c92a46 100644
--- a/net/minecraft/server/network/ServerCommonPacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerCommonPacketListenerImpl.java
@@ -29,6 +29,7 @@ import net.minecraft.util.VisibleForDebug;
 import net.minecraft.util.profiling.Profiler;
 import org.jspecify.annotations.Nullable;
 import org.slf4j.Logger;
+import dev.pulsemc.pulse.network.PulseBuffer;
 
 public abstract class ServerCommonPacketListenerImpl implements ServerCommonPacketListener {
     private static final Logger LOGGER = LogUtils.getLogger();
@@ -62,6 +63,7 @@ public abstract class ServerCommonPacketListenerImpl implements ServerCommonPack
     public @Nullable String playerBrand;
     public final java.util.Set<String> pluginMessagerChannels;
     // Paper end - retain certain values
+    public PulseBuffer pulseBuffer; // Pulse - Embedding PulseBuffer
 
     public ServerCommonPacketListenerImpl(MinecraftServer server, Connection connection, CommonListenerCookie cookie) {
         this.server = server;
@@ -75,6 +77,7 @@ public abstract class ServerCommonPacketListenerImpl implements ServerCommonPack
         this.pluginMessagerChannels = cookie.channels();
         this.keepAlive = cookie.keepAlive();
         // Paper end
+        this.pulseBuffer = new PulseBuffer(this); // Pulse - Embedding PulseBuffer
     }
 
     // Paper start - configuration phase API
@@ -321,29 +324,31 @@ public abstract class ServerCommonPacketListenerImpl implements ServerCommonPack
         this.send(packet, null);
     }
 
+    // Pulse start - Embedding PulseBuffer
     public void send(Packet<?> packet, @Nullable ChannelFutureListener sendListener) {
-        // CraftBukkit start
-        if (packet == null || this.processedDisconnect) { // Spigot
+        if (packet == null || this.processedDisconnect) return;
+        if (this.pulseBuffer != null) {
+            this.pulseBuffer.add(packet, sendListener);
+            // If packet close connection send it as "Instant"
+            if (packet.isTerminal()) {
+                this.pulseBuffer.flush();
+                this.close();
+            }
             return;
-        } else if (packet instanceof net.minecraft.network.protocol.game.ClientboundSetDefaultSpawnPositionPacket defaultSpawnPositionPacket && this instanceof ServerGamePacketListenerImpl serverGamePacketListener) {
-            serverGamePacketListener.player.compassTarget = org.bukkit.craftbukkit.util.CraftLocation.toBukkit(defaultSpawnPositionPacket.respawnData().pos(), serverGamePacketListener.getPlayer().level());
-        }
-        // CraftBukkit end
-        if (packet.isTerminal()) {
-            this.close();
         }
-
-        boolean flag = !this.suspendFlushingOnServerThread || !this.server.isSameThread();
+            // Fallback if buffer doesnt created
+            if (packet instanceof net.minecraft.network.protocol.game.ClientboundSetDefaultSpawnPositionPacket p && this instanceof ServerGamePacketListenerImpl s) {
+                s.player.compassTarget = org.bukkit.craftbukkit.util.CraftLocation.toBukkit(p.respawnData().pos(), s.getPlayer().level());
+            }
+            if (packet.isTerminal()) this.close();
 
         try {
-            this.connection.send(packet, sendListener, flag);
+            this.connection.send(packet, sendListener, true);
         } catch (Throwable var7) {
-            CrashReport crashReport = CrashReport.forThrowable(var7, "Sending packet");
-            CrashReportCategory crashReportCategory = crashReport.addCategory("Packet being sent");
-            crashReportCategory.setDetail("Packet class", () -> packet.getClass().getCanonicalName());
-            throw new ReportedException(crashReport);
+            throw new net.minecraft.ReportedException(net.minecraft.CrashReport.forThrowable(var7, "Sending packet"));
         }
     }
+    // Pulse end - Embedding PulseBuffer
 
     @Deprecated @io.papermc.paper.annotation.DoNotUse // Paper - kick event causes
     public void disconnect(Component reason) {
diff --git a/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index 5d34a8f0eae2248ba878fdcab7ad5302282dfda2..866ccc8f5376e7ddce73fe3d6480e2180391ea92 100644
--- a/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -384,6 +384,11 @@ public class ServerGamePacketListenerImpl
                 }
                 // Purpur end - AFK API
                 this.disconnect(Component.translatable("multiplayer.disconnect.idling"), org.bukkit.event.player.PlayerKickEvent.Cause.IDLING); // Paper - kick event cause
+                // Pulse start - Add buffer flush
+                if (this.pulseBuffer != null) {
+                    this.pulseBuffer.flush();
+                }
+                // Pulse end - Add buffer flush
             }
         }
     }
